packages/esm-patient-chart-app/src/workspace/workspace-window.component.tsx

```diff
@@ -44,6 +59,44 @@ const WorkspaceWindow: React.FC<ContextWorkspaceParams> = () => {
     closeWorkspace = () => {},
   } = useMemo(() => workspaces?.[0] ?? ({} as OpenWorkspace), [workspaces]);
 
+  useEffect(() => {
+    const handleRouting = (event) => {
+      const {
+        detail: { cancelNavigation, newUrl },
+      } = event as { detail: { cancelNavigation: () => void; newUrl: string } };
+
+      const regex = new RegExp(/\/patient\/([a-zA-Z0-9\-]+)\/?/);
+      const isNewUrlPatientChartUrl = regex.test(newUrl);
+      const canCloseAllWorkspaces = getWorkspaceStore()
+        .getState()
+        .openWorkspaces.every(({ name }) => {
+          const canCloseWorkspace = getWhetherWorkspaceCanBeClosed(name);
+          return canCloseWorkspace;
+        });
+
+      if (!isNewUrlPatientChartUrl) {
+        if (!canCloseAllWorkspaces) {
+          cancelNavigation();
+          const navigateToNewUrl = () => {
+            function getUrlWithoutPrefix(url: string) {
+              return url.split(window['getOpenmrsSpaBase']())?.[1];
+            }
+            navigate({ to: `\${openmrsSpaBase}/${getUrlWithoutPrefix(newUrl)}` });
+          };
+
+          closeAllWorkspaces(navigateToNewUrl);
```

Comment:
I referred [this](https://github.com/openmrs/openmrs-esm-patient-management/blob/c3d3d5e5780e605906f8052abbeddab9af3cb46d/packages/esm-patient-registration-app/src/patient-registration/before-save-prompt.tsx#L66) in the patient registration from Florian.

The `newUrl` contains the domain as well, hence it might not be the same handling by single-spa (navigating to /home vs navigating to `dev3.openmrs.org/openmrs/spa/home`). Navigate function calls the single-spa's navigation.